File: backend/app/schemas/benchmark.py
"""
Phase 29 — Forecasting Model Benchmarking Framework
Pydantic schemas for model evaluation results, comparison reports, and metrics.
"""
from typing import List, Optional, Dict, Any
from pydantic import BaseModel


class ModelMetrics(BaseModel):
    """Evaluation metrics for a single model on a single dataset."""
    model_name: str
    dataset_name: str
    metric_type: str            # power | traffic | water | internet

    # ── Accuracy metrics ────────────────────────────────────────────────
    mae: float                  # Mean Absolute Error
    rmse: float                 # Root Mean Square Error
    mape: float                 # Mean Absolute Percentage Error (%)
    r2_score: float             # Coefficient of Determination

    # ── Speed metrics ───────────────────────────────────────────────────
    training_time_ms: float
    inference_time_ms: float
    total_time_ms: float

    # ── Additional diagnostics ──────────────────────────────────────────
    n_train: int                # Training set size
    n_test: int                 # Test set size
    horizon_steps: int          # Forecast horizon (steps ahead)
    converged: bool             # Model converged without error
    error_message: Optional[str]


class BenchmarkDataset(BaseModel):
    """Metadata for a single benchmark dataset."""
    dataset_name: str
    metric_type: str
    n_samples: int
    train_size: int
    test_size: int
    seasonality_period: int
    trend_slope: float
    noise_std: float
    description: str


class ModelComparison(BaseModel):
    """
    Head-to-head comparison of two models on a specific metric.
    """
    winner: str                 # model name that won
    loser: str
    metric_name: str            # mae | rmse | mape
    winner_value: float
    loser_value: float
    improvement_pct: float      # % improvement winner has over loser
    is_significant: bool        # |improvement| > 5%


class BenchmarkRun(BaseModel):
    """
    Results for a single benchmark run across all models and datasets.
    """
    run_id: str
    timestamp: str
    models_evaluated: List[str]
    datasets_evaluated: List[str]

    # ── Per-model metrics (model_name → list of per-dataset metrics) ──
    results: List[ModelMetrics]

    # ── Aggregate metrics (mean across datasets) ───────────────────────
    aggregate: Dict[str, Dict[str, float]]   # model → {mae, rmse, mape, train_ms, infer_ms}

    # ── Head-to-head comparisons ──────────────────────────────────────
    comparisons: List[ModelComparison]

    # ── Rankings ──────────────────────────────────────────────────────
    ranking_by_mae: List[str]   # sorted model names, best first
    ranking_by_rmse: List[str]
    ranking_by_mape: List[str]
    ranking_by_speed: List[str] # fastest first

    overall_winner: str         # best across all criteria
    overall_winner_reason: str

    # ── Evaluation report ─────────────────────────────────────────────
    report_summary: str         # Multi-line executive summary
    recommendations: List[str]
    total_benchmark_time_ms: float


class BenchmarkRequest(BaseModel):
    metric_types: List[str] = ["power", "traffic", "water", "internet"]
    horizon_steps: int = 24
    n_samples: int = 200
    include_ets: bool = True    # Include Exponential Smoothing as 3rd model

File: frontend/src/api/benchmark.ts
import { getWithRetry, postWithRetry } from "./client";

// ==============================================================================
// Response Types
// ==============================================================================

export interface ModelMetrics {
  model_name: string;
  dataset_name: string;
  metric_type: string;
  mae: number;
  rmse: number;
  mape: number;
  r2_score: number;
  training_time_ms: number;
  inference_time_ms: number;
  total_time_ms: number;
  n_train: number;
  n_test: number;
  horizon_steps: number;
  converged: boolean;
  error_message: string | null;
}

export interface ModelComparison {
  winner: string;
  loser: string;
  metric_name: string;
  winner_value: number;
  loser_value: number;
  improvement_pct: number;
  is_significant: boolean;
}

export interface BenchmarkRun {
  run_id: string;
  timestamp: string;
  models_evaluated: string[];
  datasets_evaluated: string[];
  results: ModelMetrics[];
  aggregate: Record<string, Record<string, number>>;
  comparisons: ModelComparison[];
  ranking_by_mae: string[];
  ranking_by_rmse: string[];
  ranking_by_mape: string[];
  ranking_by_speed: string[];
  overall_winner: string;
  overall_winner_reason: string;
  report_summary: string;
  recommendations: string[];
  total_benchmark_time_ms: number;
}

export interface DatasetPreview {
  metric_type: string;
  description: string;
  n_samples: number;
  train_size: number;
  test_size: number;
  seasonality_period: number;
  values: number[];
  train_values: number[];
  test_values: number[];
  stats: { mean: number; std: number; min: number; max: number };
}

export interface BenchmarkRequest {
  metric_types: string[];
  horizon_steps: number;
  n_samples: number;
  include_ets: boolean;
}

// ==============================================================================
// API Functions
// ==============================================================================

export async function runBenchmark(req: BenchmarkRequest): Promise<BenchmarkRun> {
  return postWithRetry<BenchmarkRun>("/api/v1/benchmark/run", req);
}

export async function runQuickBenchmark(
  metric_type = "power",
  n_samples = 120,
  horizon = 12
): Promise<BenchmarkRun> {
  return getWithRetry<BenchmarkRun>(
    `/api/v1/benchmark/quick?metric_type=${metric_type}&n_samples=${n_samples}&horizon=${horizon}`
  );
}

export async function fetchDatasetPreview(
  metric_type: string,
  n_samples = 200
): Promise<DatasetPreview> {
  return getWithRetry<DatasetPreview>(
    `/api/v1/benchmark/preview/${metric_type}?n_samples=${n_samples}`
  );
}

// ==============================================================================
// Colour palette (shared with component)
// ==============================================================================
export const MODEL_COLORS: Record<string, string> = {
  Prophet: "#a855f7",
  ARIMA:   "#00e5ff",
  ETS:     "#22c55e",
};

export const MODEL_ICONS: Record<string, string> = {
  Prophet: "🔮",
  ARIMA:   "📈",
  ETS:     "📊",
};

export const METRIC_LABELS: Record<string, string> = {
  mae:  "MAE",
  rmse: "RMSE",
  mape: "MAPE (%)",
  r2:   "R² Score",
};
